### PR TITLE
feat: custom error file

### PR DESCRIPTION
**Pull request by AdeosunDeji**

### CODE REVIEW

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:1]: I see you've done your research in the code, excellent! Some great suggestions in here.

Regarding the structure I could have been more explicit in the ticket. I would like the custom errors to extend the Error base class.

Furthermore we should create a folder called error under domain. Then move `error.ts` into `domain/error/error.ts`. 

Perhaps we should even rename `error.ts` to `sleepapi-error.ts`and add in a base Sleep API error class like:
```typescript
export class SleepAPIError extends Error {
  static message(message: string) {
    return new this(message);
  }
}
```

After that we can create folders under `domain/error` to group different error types. For example we could have `domain/error/database-error` and move `DatabaseMigrationError`, `DatabaseConnectionError` etc in there.

Instead of using const variables for the errors I would like them to extend this `SleepAPIError` base error class.
So it would look something like:
```typescript
export class DatabaseMigrationError extends SleepAPIError {}
```

This should allow us to re-use the error types without limiting which message we include as the base `SleepAPIError` would allow for passing custom messages.

For example `DatabaseMigrationError` could be used both for incorrect migration type as you suggest right now, or if we crash during migration due to a schema issue etc.

This also means we're mostly looking to setup the types here rather than the messages. The message we'll pass just like we're passing them to the normal `Error` base class as we do today.

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:3]: To align with your other naming standard suggestions I would rename the following database errors:
`EntryNotFoundError` to `DatabaseNotFoundError`
`InsertError` to `DatabaseInsertError`

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:6]: As per my previous suggestion of creating common re-usable error types this would not be needed.
Instead we'll throw a `DatabaseNotFoundError` and pass in the meal-specific message.

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:7]: Regarding the following errors:
- `Level60Error`
- `EnergyPercentageError`
- `LevelError`
- `NatureError`

These could all be replaced by an error called for example `StatError` and be moved into `domain/error/stat/stat-error.ts`

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:12]: This could be renamed to just `PokemonError` and be moved into `domain/error/pokemon/pokemon-error.ts`

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:14]: This `SubTeamsError`can be removed since this indicates a ProgrammingError which you already have added

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:15]: This could be moved into `domain/error/ingredient/ingredient-error.ts`

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:16]: I would like to rename `ProgrammerError` to `ProgrammingError` as it suggest an issue with the code rather than blaming an individual programmer. This can be moved into `domain/error/programming/programming-error.ts`

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:18]: We should rename this to `MockError` and perhaps we can create a test folder under domain for this so we may add in more test-related errors. So move the `MockError` into `domain/error/test/mock-error.ts`

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:19]: How about something more generic like `ConfigError` and move it into `domain/error/config/config-error.ts`

File: backend/src/domain/error.ts
Comment (SleepAPI) [inline backend/src/domain/error.ts:20]: This could be renamed to `MealError` and be moved into `domain/error/meal/meal-error.ts`

**Review by SleepAPI — CHANGES_REQUESTED** [review]
It looks like your files use CRLF line ending so I'm assuming you're developing on a Windows machine. This causes big diffs for all files that you saved on your machine.
This also makes the PR difficult to review and if we allowed formatting changes between different platforms then every other PR would have this issue.

We have explicitly added `"files.eol": "\n"` to the `sleepapi.code-workspace` to fix this.
So this suggests to me that likely you either didn't work in Visual Studio Code or didn't use the workspace. Please read our [DEVELOPMENT_SETUP](https://github.com/SleepAPI/SleepAPI/blob/develop/DEVELOPMENT_SETUP.md) if you need help with this.

Furthermore it seems that you removed some line endings. Prettier is set as the default autoformat in the workspace and prettier is configured by default to add newlines at the end of the file. Prettier also formats LF as line endings by default. This also suggests to me that VS Code and/or the workspace was not used.

I noticed our .editorconfig states we should not add new lines at the end of a file, but this is unintended and as previously mentioned gets overridden by prettier.

Final note; great suggestions for different error types. You've clearly gone through the code and summarized which errors we throw today, excellent work! 

I left some comments on how I think we should structure the errors. 
Please fix those and the LF/newline formatting 😃

File: backend/src/domain/error.ts
Comment (AdeosunDeji) [reply backend/src/domain/error.ts:1]: Thanks for the corrections and guidance. Your insight is greatly appreciated.

**Review by AdeosunDeji — COMMENTED** [review]
Made changes. Waiting for your feedback.

**Review by SleepAPI — CHANGES_REQUESTED** [review]
Big improvement. Left a few comments and I have a few more comments here.

Currently you implement each error class like so:
```typescript
export class DatabaseMigrationError extends SleepAPIError {
  constructor(message = 'DATABASE_MIGRATION is optional, but if set must be one of [UP, DOWN]') {
    super(message);
    this.name = 'DatabaseMigrationError';
  }
}
```

I would prefer it to be completely open types instead like:
```typescript
export class DatabaseMigrationError extends SleepAPIError {}
```

That way you can throw the same error in many places, but pass different messages, making it more generic. I would like to avoid case-specific message in places where we're trying to be generic. Please update all error types to open objects.

It also looks like you're still writing the files with CRLF instead of LF. Please see my comment on the previous commit detailing how to correctly format the files with VS Code and the workspace.

Finally I see you're adding new commits to the PR, that's perfectly fine, but since we're just fixing the commits introduced on this PR I would like the final result of them to be just one commit. So either you can squash those 2 commits now and then amend any future fixes or keep pushing new commits and then in the end squash them all together before I merge.

Please let me know if you have any questions 👍

**Pull request closed without merging**
